### ISSUE TITLE
Algorithm returns full path as field name [Windows 7]

### ISSUE THREAD

**Issue by eivindhammers** [jannes-m/RQGIS#39]
I use RQGIS to perform zonal statistics with SAGA, via QGIS (2.14.10 on Windows 7 via OSGEO4W). However, the output returns the beginning of the path as field name of the statistic (C.Userseivi in the example below). If I'm not doing something wrong, I think `processing.runalg('saga:gridstatisticsforpolygons')` (and probably many other algs) automatically names the output field(s) like "pathtoraster [fun]", which would be "C:/Users/[...]temp/TEMPDIR/GRID.asc [sum]" in the example below.

Would it be an idea to optionally specify the output field name in `run_qgis`? That would be setting the name of the tempfile to something other than "GRID.asc", and then not provide the full path to that file in "py_cmd.py", instead relying on setting the working directory.

```R
qgis <- set_env(root = "C:/OSGeo4W64/")
alg <- "saga:gridstatisticsforpolygons"
params <- get_args_man(alg, qgis_env = qgis)
params$GRIDS <- pop.afr # RasterLayer
params$POLYGONS <- grid.sp # SPDF
params$SUM <- "TRUE"
exclude <- c("COUNT", "MIN", "MAX", "RANGE", "MEAN", "VAR", "STDDEV")
params[exclude] <- "FALSE"
params$RESULT <- "output.shp"
grid.pop <- run_qgis(alg, params = params, qgis_env = qgis, 
                     load_output = params$RESULT)
# grid.pop now includes a column "C.Userseivi"
```

**Comment by jannes-m** [comment]
Hi Eivind,

first of all, thank you for using and testing `RQGIS`! Let me first point out that it would be safer to use Python notation instead of R syntax (maybe we should take care of this as well...):
```r 
params$SUM <- "True"
params[exclude] <- "False"
```

Secondly, it is also safer to indicate the full path as output:

```r
params$RESULT <- file.path(tempdir(), "output.shp")
```

Ok, but that's not what you wanted to know. I see what you mean. Honestly, I have never used these zonal statistical functions (always used R for it :)). Therefore, let us first think about your suggestion and its possible implications/side-effects. But maybe your suggestion is a bit difficult to implement since we want to leave the choice to the user to specify the path to a specific folder where to save the output shapefile...

**Comment by eivindhammers** [comment]
Thanks for quick feedback. You're right about the python booleans, of course.

Regarding full path as output, I'm following the `run_qgis` example here, where only "output.shp" is specified. Not sure what you mean by "safer"; when I specify `load_output = x`, I normally don't need the "output.shp" in my project folder, since I will continue working with the data in R, so I think having RQGIS handling tempfiles and tempdirs in the background is better (unless I leave `load_output = NULL`).

The reason for using zonal statistics in QGIS is that `raster::extract()` is currently painfully slow when the raster and/or vector data is big, although I know that `raster` is about to be updated.

Note that the location of "output.shp" does not matter for field names, but the input raster does. E.g. manually changing "py_cmd.py", setting the input raster to "GRID.asc" (which works, since it resides in the same folder as the .py) gives an output field name of something like "GRIDasc".

Any solution in RQGIS is suboptimal as long as `runalg('saga:gridstatisticsforpolygons')` does not allow specifying output field names. But perhaps renaming the temporary file "GRID.asc" to something containing the name of the `RasterLayer` makes sense? Fields will be truncated to 10 characters, though..

**Comment by pat-s** [comment]
> The reason for using zonal statistics in QGIS is that raster::extract() is currently painfully slow when the raster and/or vector data is big, although I know that raster is about to be updated.

Try processing in chunks/block as suggested in the `raster` vignettes. You can even use a cluster for single processing steps like `raster::extract()`. I assume that this way should be more efficient than using SAGA algorithms through RQGIS. 

Thanks for your feedback regarding the field naming issue, we should definitely think about this and address it in a proper way!

**Comment by eivindhammers** [comment]
Thanks for the tip. It turns out SAGA is way faster than `raster::extract()` even when running in parallel, by a factor of about 60, which seems too much. I'll investigate, but in the meantime I'm sticking to RQGIS + SAGA.

**Comment by pat-s** [comment]
> SAGA is way faster than raster::extract() even when running in parallel, by a factor of about 60

Is that based on your own experiments or is there any source comparing both? Just curious - factor of 60 is a tremendous difference. However that's off-topic here - maybe you can send me more information via mail (check my website).

**Comment by eivindhammers** [comment]
My own experience; I ran it in parallel following your suggestion on three local nodes (cores), and it took 53 minutes, compared to less than 1 minute with `saga:gridstatisticsforpolygons`. I'll run it again to make sure that I've not done any mistakes.

**Comment by eivindhammers** [comment]
Re: your question in #43, I have looked into a solution, but I'm not sure how to implement. My initial idea was to get the names of all the statistics set to `"TRUE"` in `params`, prefix them with `r_` (which keeps them within the shapefile field name character limit), and add to `py_cmd.py` something like

```python
outfile = "output.shp"
os.system('ogr2ogr '+ outfile + ' ' + outfile + ' -sql \"SELECT ' + oldfield + ' AS ' + newfield + ' from outfile\"')
```

This basically copies and replaces "output.shp" with new field names. The problem is that all statistics are given the same field name in the "output.shp", in my case "C:userseivi", so I am not sure how to refer to them in the `ogr2ogr` SQL statement.

When loaded into R, the identically-named fields are given number-suffixes, most likely in the same order as the `SUM = 'TRUE', COUNT = 'TRUE'`, etc., are provided to the `processing.runalg`, so I don't think I'll spend too much time on a fix. The best solution would be to add a `prefix` optional argument to the python plugin, as in `"qgis:zonalstatistics"`, but that's beyond my capabilities.

**Comment by eivindhammers** [comment]
Afaics, this is fixed in QGIS 2.18.5.

**Comment by jannes-m** [comment]
Self-solving problems, I love it!
